extensions/ql-vscode/src/data-extensions-editor/predicates.ts

```diff
@@ -0,0 +1,136 @@
+import { ExternalApiUsage } from "./external-api-usage";
+import {
+  ModeledMethod,
+  ModeledMethodType,
+  ModeledMethodWithSignature,
+} from "./modeled-method";
+
+export type ExternalApiUsageByType = {
+  externalApiUsage: ExternalApiUsage;
+  modeledMethod: ModeledMethod;
+};
+
+export type ExtensiblePredicateDefinition = {
+  extensiblePredicate: string;
+  generateMethodDefinition: (method: ExternalApiUsageByType) => any[];
+  readModeledMethod: (row: any[]) => ModeledMethodWithSignature;
+
+  supportedKinds?: string[];
+};
+
+function readRowToMethod(row: any[]): string {
+  return `${row[0]}.${row[1]}#${row[3]}${row[4]}`;
+}
+
+export const extensiblePredicateDefinitions: Record<
+  Exclude<ModeledMethodType, "none">,
+  ExtensiblePredicateDefinition
+> = {
+  source: {
+    extensiblePredicate: "sourceModel",
+    // extensible predicate sourceModel(
+    //   string package, string type, boolean subtypes, string name, string signature, string ext,
+    //   string output, string kind, string provenance
+    // );
+    generateMethodDefinition: (method) => [
+      method.externalApiUsage.packageName,
+      method.externalApiUsage.typeName,
+      true,
+      method.externalApiUsage.methodName,
+      method.externalApiUsage.methodParameters,
+      "",
+      method.modeledMethod.output,
+      method.modeledMethod.kind,
+      "manual",
+    ],
+    readModeledMethod: (row) => ({
+      signature: readRowToMethod(row),
+      modeledMethod: {
+        type: "source",
+        input: "",
+        output: row[6],
+        kind: row[7],
+      },
+    }),
+    supportedKinds: ["remote"],
+  },
+  sink: {
+    extensiblePredicate: "sinkModel",
+    // extensible predicate sinkModel(
+    //   string package, string type, boolean subtypes, string name, string signature, string ext,
+    //   string input, string kind, string provenance
+    // );
+    generateMethodDefinition: (method) => [
+      method.externalApiUsage.packageName,
+      method.externalApiUsage.typeName,
+      true,
+      method.externalApiUsage.methodName,
+      method.externalApiUsage.methodParameters,
+      "",
+      method.modeledMethod.input,
+      method.modeledMethod.kind,
+      "manual",
+    ],
+    readModeledMethod: (row) => ({
+      signature: readRowToMethod(row),
+      modeledMethod: {
+        type: "sink",
+        input: row[6],
+        output: "",
+        kind: row[7],
+      },
+    }),
+    supportedKinds: ["sql", "xss", "logging"],
```

Comment:
This is a good start. Each language will have its own set of kinds. We will also need to support custom sink kinds. For instance, javascript uses `sql-injection`. `credentials[username]`, `credentials[password]`, `mongodb.sink`, and others.

I think csharp is the same as java, but i haven't done a deep analysis yet.

The ideal situation is that each language exposes the default sink/source kinds it supports through a contextual query and this editor requests that information before displaying. But, that is too much for this initial cut. For now, I don't think it makes sense to hard code this.

Reply:
As discussed we're only supporting Java for now, so we fully intend to change this in the future to support other languages.